src/main/java/com/autotune/analyzer/recommendations/updater/vpa/VpaUpdaterImpl.java

```diff
@@ -0,0 +1,371 @@
+ * Copyright (c) 2024 Red Hat, IBM Corporation and others.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *******************************************************************************/
+
+package com.autotune.analyzer.recommendations.updater.vpa;
+
+import com.autotune.analyzer.exceptions.ApplyRecommendationsError;
+import com.autotune.analyzer.exceptions.UnableToCreateVPAException;
+import com.autotune.analyzer.kruizeObject.KruizeObject;
+import com.autotune.analyzer.recommendations.RecommendationConfigItem;
+import com.autotune.analyzer.recommendations.updater.RecommendationUpdaterImpl;
+import com.autotune.analyzer.utils.AnalyzerConstants;
+import com.autotune.analyzer.utils.AnalyzerErrorConstants;
+import com.autotune.common.k8sObjects.K8sObject;
+import com.autotune.common.data.result.ContainerData;
+import com.autotune.analyzer.recommendations.objects.MappedRecommendationForTimestamp;
+import com.autotune.analyzer.recommendations.objects.TermRecommendations;
+import io.fabric8.autoscaling.api.model.v1.*;
+import io.fabric8.kubernetes.api.model.ObjectMeta;
+import io.fabric8.kubernetes.api.model.Quantity;
+import io.fabric8.kubernetes.api.model.apiextensions.v1.CustomResourceDefinitionList;
+import io.fabric8.kubernetes.api.model.autoscaling.v1.CrossVersionObjectReferenceBuilder;
+import io.fabric8.kubernetes.client.DefaultKubernetesClient;
+import io.fabric8.kubernetes.client.KubernetesClient;
+import io.fabric8.kubernetes.client.dsl.ApiextensionsAPIGroupDSL;
+import io.fabric8.verticalpodautoscaler.client.DefaultVerticalPodAutoscalerClient;
+import io.fabric8.verticalpodautoscaler.client.NamespacedVerticalPodAutoscalerClient;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.sql.Timestamp;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+public class VpaUpdaterImpl extends RecommendationUpdaterImpl {
+    private static final Logger LOGGER = LoggerFactory.getLogger(VpaUpdaterImpl.class);
+    private static VpaUpdaterImpl vpaUpdater = new VpaUpdaterImpl();
+
+    private KubernetesClient kubernetesClient;
+    private ApiextensionsAPIGroupDSL apiextensionsClient;
+
+
+    private VpaUpdaterImpl() {
+        this.kubernetesClient = new DefaultKubernetesClient();
+        this.apiextensionsClient = kubernetesClient.apiextensions();
+    }
+
+    public static VpaUpdaterImpl getInstance() {
```

Comment:
To address this, I've updated the `getInstance` method to be `synchronized` method, ensuring thread safety. And yes, this method is part of the flow for updater service as you assumed.

Reply:
Thanks for the changes @shekhar316 

After taking a look at the `synchronized` method. I feel my second suggestion (To add a `synchronized` block) might be more helpful than the first one (To make method `synchronized`)

Here is my analysis and explanation:

Let's say we have two parallel requests are being processed by two threads of web server which will call `getInstance` method. In this case due to `synchronized` modifier only one thread gets the access and creates the object, meanwhile the other thread waits till the first thread completes the object creation.

But the problem now is the subsequent requests processed by the threads will wait till the instance is returned, while post object creation we can allow parallel threads to make use of object without any wait.

So as the object creation is the only costly operation we would like to avoid threads doing it in parallel, Let's place a check for `null` and then add `synchronized` block and if the object exists we will just return it.

Existing code:

```java
public static synchronized VpaUpdaterImpl getInstance() {
    if (vpaUpdater == null) {
        vpaUpdater = new VpaUpdaterImpl();
    }
    return vpaUpdater;
}
```

After Modification:

```java
public static VpaUpdaterImpl getInstance() {
    // Return object if not null
    if (null != vpaUpdater)
        return vpaUpdater;

    // Allow object creation only in a synchronized block
    synchronized (VpaUpdaterImpl.class) {
        // Add one more check for null as at the time of first creation,
        // two threads might be waiting for sync block access and one 
        // would have been waiting in an assumption that object is null 
        // but the earlier thread would have created the object so now once
        // the other thread enters the block it should check for null again.
        if (null == vpaUpdater) {
            vpaUpdater = new VpaUpdaterImpl();
        }
    }
    return vpaUpdater;
}
```

I'm more inclined to have a synchronized block instead of having a synchronized method which might add latency for threads to work in parallel, post creation of the object.
 
@shekhar316  Please let me know your thoughts on this approach. Thanks!

Reply:
Thanks @bharathappali for your detailed analysis and explanation! I agree with your thought. The `synchronized` block with double-checking is a better approach compared to synchronizing the entire method. This addresses the performance concerns as well as ensures thread safety during object creation.

According to the approach, I have modified the function to use a `synchronized` block rather than `synchronized` function.